File: utils/rra/rra2wav.c
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include "/usr/local/include/songlib/util.h"
#include "/usr/local/include/songlib/rra.h"
#include "/usr/local/include/songlib/wave.h"

/* change PROGRAM_NAME and PROGRAM_VERSION appropriately */

char *PROGRAM_NAME = "rra2wav";
char *PROGRAM_VERSION = "0.01";

static double Amplitude = 1;
static double Skip = 0;

static void filter(FILE *,FILE *);
static void batch(int,char **);
static int processOptions(int, char **);

int
main(int argc,char **argv)
    {
    int argIndex = 1;

    argIndex = processOptions(argc,argv);

    if (argc-argIndex == 0)
        {
        filter(stdin,stdout);
        }
    else if (argc-argIndex == 1)
        {
        filter(OpenFile(argv[argIndex],"r"),stdout);
        }
    else if (argc-argIndex >= 2 && strstr(argv[argIndex+1],".rra") != 0)
        {
        batch(argc-argIndex,argv+argIndex);
        }
    else if (argc-argIndex == 2)
        {
        filter(OpenFile(argv[argIndex],"r"),OpenFile(argv[argIndex+1],"w"));
        }
    else
        {
        printf("usage: rra2wav [<input rra file> [<output wav file>]]\n");
        printf("       rra2wav <input rra file1> <input rra file2> ...\n");
        exit(-1);
        }

    return 0;
    }

static void
filter(FILE *in,FILE *out)
    {
    int i,j,amp;
    int skipSamples;
    RRA *h;

    h = newRRAHeader();
    
    readRRAHeader(in,h,0);

    skipSamples = Skip * h->sampleRate;

    writeWaveHeader(out,h->sampleRate,h->channels,
        h->bitsPerSample,h->samples - h->skip - skipSamples);

    amp = readRRAAmplitude(in,out,h->bitsPerSample,discardComment);
    for (i = 0; !feof(in) && i < h->samples; ++i)
        for (j = 0; !feof(in) && j < h->channels; ++j)
            {
            if (i >= (h->skip + skipSamples) * h->channels)
                writeByteInt(out,amp * Amplitude,h->bitsPerSample/8);
            amp = readRRAAmplitude(in,out,h->bitsPerSample,discardComment);
            }

    fclose(in);
    fclose(out);
    }

static void
batch(int argc,char **argv)
    {
    int i;
    char buffer[512];
    char *base;
    char *suffix = ".wav";
    FILE *in,*out;

    for (i = 0; i < argc; ++i)
        {
        fprintf(stderr,"converting %s to WAV\n",argv[i]);
        in = OpenFile(argv[i],"r");
        base = strtok(argv[i],".");
        if (strlen(base) > sizeof(buffer) - (strlen(suffix) + 2))
            Fatal("file name too long\n");
        sprintf(buffer,"%s%s",base,suffix);
        out = OpenFile(buffer,"w");
        filter(in,out);
        }
    }

/* only -oXXX  or -o XXX options */

static int
processOptions(int argc, char **argv)
{
    int argIndex;
    int argUsed;
    int separateArg;
    char *arg;

    argIndex = 1;

    while (argIndex < argc && *argv[argIndex] == '-') {

        separateArg = 0;
        argUsed = 0;

        if (argv[argIndex][2] == '\0')
            {
            arg = argv[argIndex+1];
            separateArg = 1;
            }
        else
            arg = argv[argIndex]+2;

        switch (argv[argIndex][1])
            {
            /*
             * when option has an argument, do this
             *
             *     examples are -m4096 or -m 4096
             *
             *     case 'm':
             *         MemorySize = atol(arg);
             *         argUsed = 1;
             *         break;
             *
             *
             * when option does not have an argument, do this
             *
             *     example is -a
             *
             *     case 'a':
             *         PrintActions = 1;
             *         break;
             */
            case 'a':
                Amplitude = atof(arg);
                argUsed = 1;
                break;
            case 's':
                Skip = atof(arg);
                argUsed = 1;
                break;
            case 'v':
                printf("%s version %s\n", PROGRAM_NAME, PROGRAM_VERSION);
                exit(0);
                break;
            default:
                Fatal("option %s not understood\n",argv[argIndex]);
            }

        if (separateArg && argUsed)
            ++argIndex;

        ++argIndex;
        }

    return argIndex;
}

File: utils/rra/rra2wav.py
import sys

# change PROGRAM_NAME and PROGRAM_VERSION appropriately

PROGRAM_NAME = "rra2wav (simplified python3 version)"
PROGRAM_VERSION = "0.01"

def main():
    if (len(sys.argv) == 1):
        s = sys.stdin
    elif (len(sys.argv) == 2):
        s = open(sys.argv[1],"r")
    else:
        print("usage: python3 rra2wav.py RRA_FILENAME > WAV_FILENAME")
        print("   or: python3 rra2wav.py < RRA_FILENAME > WAV_FILENAME")
        print
        sys.exit(1)

    filter(s)
    s.close()
    return 0

def filter(s):

    sr,ch,bps,samples,amp = readRRAHeader(s)

    writeWaveHeader(sr,ch,bps,samples)

    for i in range(0,samples,1):
        for j in range(0,ch,1):
            printByteInt(amp,bps//8)
            amp = readint(s)

def readRRAHeader(s):

    # default RRA values
    sampleRate = 44100
    bps = 16
    channels = 1
    samples = 10 * sampleRate # default is 10 seconds worth of samples

    attribute = readtoken(s)

    if (attribute == ""):
        print("not an RRA file: first attribute should be RRAUDIO")
        sys.exit(1)

    # check if raw data file -- if so, just return
    if (attribute[0].isdigit()):
        return sampleRate,channels,bps,samples,int(attribute) #attribute is a sample

    if (attribute != "RRAUDIO"):
        print("not an RRA file: first attribute is",attribute,"(should be RRAUDIO)")
        sys.exit(1)

    attribute = readtoken(s)
    while (attribute != "%%"):

        value = readtoken(s)

        if (attribute == "samples:"):
            samples = int(value)
        elif (attribute == "sampleRate:"):
            sampleRate = int(value)
        elif (attribute == "bitsPerSample:"):
            bps = int(value)
        elif (attribute == "channels:"):
            channels = int(value)
        #else:
            #print("ignoring RRA attribute:",attribute,file=sys.stderr)

        attribute = readtoken(s)

    amp = readint(s)
    return sampleRate,channels,bps,samples,amp #attribute is a sample

def writeWaveHeader(rate,channels,bps,samples):
    sys.stdout.buffer.write(b"RIFF")
    printByteInt(36 + samples * channels * (bps // 8),4)
    sys.stdout.buffer.write(b"WAVE")
    sys.stdout.buffer.write(b"fmt ")
    printByteInt(16,4) #PCM subchunk
    printByteInt(1,2) #PCM (no compression)
    printByteInt(channels,2)  #1 for mono, 2 for stereo
    printByteInt(rate,4) #sample rate
    printByteInt(rate * channels * (bps // 8),4) #byte rate
    printByteInt(channels * (bps // 8),2) #block align
    printByteInt(bps,2) #bits per sample
    sys.stdout.buffer.write(b"data")
    printByteInt(samples * channels * (bps // 8),4) #subchunk 2 size

def printByteInt(n,size):

    b = [None]*size
    for i in range(0,size):
       b[i] = n % 256
       n = n // 256

    b = bytes(b)
    #print("b is",b,file=sys.stderr)

    sys.stdout.buffer.write(b)
       
tokens = []
tokenIndex = 0

def readtoken(fp):
    global tokens,tokenIndex
    if (tokenIndex == len(tokens)):
        tokens=fp.readline().split()
        if (len(tokens) == 0): return ""
        tokenIndex = 0
    tokenIndex += 1
    return tokens[tokenIndex-1]

def readint(fp):
    t = readtoken(fp)
    if (t != ""):
        t = int(t)
    return t

main()
